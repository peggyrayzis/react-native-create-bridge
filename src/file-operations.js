const path = require("path");
const fs = require("mz/fs");
const compareVersions = require("compare-versions");
const semver = require("../utils/semver");

const pkg = require(path.join(process.cwd(), "package.json"));

function getFileNames(dirPath) {
  return fs.readdir(dirPath).catch(e => console.error("[getFileNames] ", e));
}

function readFile(file, readDirPath) {
  return fs
    .readFile(path.join(readDirPath, file), "utf-8")
    .catch(e => console.error("[readFile] ", e));
}

function parseFile(fileData, { templateName, packageName, app, rnVersion }) {
  let kotlinPackage;
  let javaPackage;
  let version;

  try {
    version = semver(rnVersion);
  } catch (e) {
    version = "0.0.0";
  }

  // TODO: figure out a better way to handle one off breaking changes
  if (rnVersion && compareVersions(version, "0.47.2") < 0) {
    kotlinPackage = `
    override fun createJSModules(): List<Class<out JavaScriptModule>> {
        return emptyList()
    }

    `;

    javaPackage = `
    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    `;
  } else {
    kotlinPackage = "";
    javaPackage = "";
  }

  return fileData
    .replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app)
    .replace(/{{kotlinPackage}}/g, kotlinPackage)
    .replace(/{{javaPackage}}/g, javaPackage);
}

function readAndWriteFiles(files, paths, config) {
  const { readDirPath, writeDirPath } = paths;
  return Promise.all(
    files.map(file => {
      readFile(file, readDirPath).then(fileData => {
        const parsedFile = parseFile(fileData, config);
        const fileName = file.replace("Template", config.templateName);
        return fs.writeFile(path.join(writeDirPath, fileName), parsedFile);
      });
    })
  ).catch(e => console.error("[readAndWriteFiles] ", e));
}

module.exports = {
  pkg,
  getFileNames,
  readFile,
  parseFile,
  readAndWriteFiles
};
