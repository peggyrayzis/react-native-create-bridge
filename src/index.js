#! /usr/bin/env node

import inquirer from "inquirer";
import path from "path";
import isValid from "is-valid-path";
import mkdir from "mkdirp-promise";
import { success as successIcon, error as errorIcon } from "log-symbols";

import readAndWriteFiles, { pkg, getFileNames } from "./file-operations";

const templateNameRegex = /\w+/;
const promptConfig = [
  {
    type: "input",
    name: "templateName",
    message: "What is your bridge module called?",
    default: "ExampleBridge",
    validate: input => templateNameRegex.test(input)
  },
  {
    type: "checkbox",
    name: "bridgeType",
    message: "What type of bridge would you like to create?",
    default: ["Native Module", "Native UI Component"],
    choices: ["Native Module", "Native UI Component"]
  },
  {
    type: "checkbox",
    name: "environment",
    message: "What OS & languages would you like to support?",
    default: ["Android/Java", "iOS/Objective-C"],
    choices: ["Android/Java", "Android/Kotlin", "iOS/Swift", "iOS/Objective-C"]
  },
  {
    type: "input",
    name: "jsPath",
    message: "What directory should we deliver your JS files to?",
    default: ".",
    validate: input => isValid(input)
  }
];

const environmentMap = {
  "Android/Java": createJavaEnvironment,
  "Android/Kotlin": createKotlinEnvironment,
  "iOS/Swift": createSwiftEnvironment,
  "iOS/Objective-C": createObjCEnvironment
};

async function init() {
  try {
    const {
      environment,
      bridgeType,
      templateName,
      jsPath
    } = await inquirer.prompt(promptConfig);

    const templateFolder = bridgeType.length > 1
      ? "combined"
      : bridgeType[0] === "Native Module" ? "modules" : "ui-components";

    const promises = environment.map(env =>
      environmentMap[env](templateName, templateFolder)
    );

    promises.push(createJSEnvironment(templateName, templateFolder, jsPath));
    await Promise.all(promises);

    console.log(
      `${successIcon} Your bridge module was successfully created! ðŸŽ‰`
    );
  } catch (e) {
    console.log(
      `${errorIcon} Oh no! ðŸ’©  Something went wrong with creating your bridge module.\nPlease report any errors here: https://github.com/peggyrayzis/react-native-create-bridge/issues\n\nError: ${e}`
    );
  }
}

async function createJavaEnvironment(templateName, templateFolder) {
  const appPath = path.join(
    process.cwd(),
    "android",
    "app",
    "src",
    "main",
    "java",
    "com",
    pkg.name
  );

  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "android-java"
  );

  const writeDirPath = await mkdir(
    path.join(appPath, templateName.toLowerCase())
  );

  const paths = {
    readDirPath,
    writeDirPath: writeDirPath
      ? writeDirPath
      : path.join(appPath, templateName.toLowerCase())
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(
    files,
    paths,
    templateName,
    templateName.toLowerCase(),
    pkg.name.toLowerCase()
  );
}

async function createKotlinEnvironment(templateName, templateFolder) {
  const appPath = path.join(
    process.cwd(),
    "android",
    "app",
    "src",
    "main",
    "java",
    "com",
    pkg.name
  );

  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "android-kotlin"
  );

  const writeDirPath = await mkdir(
    path.join(appPath, templateName.toLowerCase())
  );

  const paths = {
    readDirPath,
    writeDirPath: writeDirPath
      ? writeDirPath
      : path.join(appPath, templateName.toLowerCase())
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(
    files,
    paths,
    templateName,
    templateName.toLowerCase(),
    pkg.name.toLowerCase()
  );
}

async function createSwiftEnvironment(templateName, templateFolder) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "ios-swift"
  );

  const paths = {
    readDirPath,
    writeDirPath: path.join(process.cwd(), "ios")
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

async function createObjCEnvironment(templateName, templateFolder) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "ios-objc"
  );

  const paths = {
    readDirPath,
    writeDirPath: path.join(process.cwd(), "ios")
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

async function createJSEnvironment(templateName, templateFolder, jsPath) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "js"
  );

  await mkdir(jsPath);

  const paths = {
    readDirPath,
    writeDirPath: jsPath
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

init();
