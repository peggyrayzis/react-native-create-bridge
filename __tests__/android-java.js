import path from 'path';
import { readFile, parseFile } from '../src/file-operations';

describe('Android/Java: UI Components', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'ui-components',
    'android-java',
  );

  it('creates a TemplateManager.java', async () => {
    const fileData = await readFile('TemplateManager.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplatePackage.java', async () => {
    const fileData = await readFile('TemplatePackage.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });
});

describe('Android/Java: Native Modules', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'modules',
    'android-java',
  );

  it('creates a TemplateModule.java', async () => {
    const fileData = await readFile('TemplateModule.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplatePackage.java', async () => {
    const fileData = await readFile('TemplatePackage.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });
});

describe('Android/Java: Combined', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'combined',
    'android-java',
  );

  it('creates a TemplateManager.java', async () => {
    const fileData = await readFile('TemplateManager.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplateModule.java', async () => {
    const fileData = await readFile('TemplateModule.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplatePackage.java', async () => {
    const fileData = await readFile('TemplatePackage.java', readDirPath);
    const parsedFile = parseFile(
      fileData,
      templateName,
      templateName.toLowerCase(),
      'testapp',
    );
    expect(parsedFile).toMatchSnapshot();
  });
});
