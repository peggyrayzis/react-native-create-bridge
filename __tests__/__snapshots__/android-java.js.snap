exports[`Android/Java: Combined creates a TemplateManager.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import android.view.View;

import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;

import com.facebook.react.uimanager.annotations.ReactProp;

public class TestModuleManager extends SimpleViewManager<View> {
    public static final String REACT_CLASS = \"TestModule\";

    @Override
    public String getName() {
        // Tell React the name of the module
        // https://facebook.github.io/react-native/docs/native-components-android.html#1-create-the-viewmanager-subclass
        return REACT_CLASS;
    }

    @Override
    public View createViewInstance(ThemedReactContext context){
        // Create a view here
        // https://facebook.github.io/react-native/docs/native-components-android.html#2-implement-method-createviewinstance
        return new View(context);
    }

    @ReactProp(name = \"exampleProp\")
    public void setExampleProp(View view, String prop) {
        // Set properties from React onto your native component via a setter method
        // https://facebook.github.io/react-native/docs/native-components-android.html#3-expose-view-property-setters-using-reactprop-or-reactpropgroup-annotation
    }
}
"
`;

exports[`Android/Java: Combined creates a TemplateModule.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import android.support.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;

public class TestModuleModule extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = \"TestModule\";
    private static ReactApplicationContext reactContext = null;

    public TestModuleModule(ReactApplicationContext context) {
        // Pass in the context to the constructor and save it so you can emit events
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        super(context);

        reactContext = context;
    }

    @Override
    public String getName() {
        // Tell React the name of the module
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        return REACT_CLASS;
    }

    @Override
    public Map<String, Object> getConstants() {
        // Export any constants to be used in your native module
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        final Map<String, Object> constants = new HashMap<>();
        constants.put(\"EXAMPLE_CONSTANT\", \"example\");

        return constants;
    }

    @ReactMethod
    public void exampleMethod () {
        // An example native method that you will expose to React
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
    }

    private static void emitDeviceEvent(String eventName, @Nullable WritableMap eventData) {
        // A method for emitting from the native side to JS
        // https://facebook.github.io/react-native/docs/native-modules-android.html#sending-events-to-javascript
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName, eventData);
    }
}
"
`;

exports[`Android/Java: Combined creates a TemplatePackage.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class TestModulePackage implements ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
      // Register your native module
      // https://facebook.github.io/react-native/docs/native-modules-android.html#register-the-module
      return Arrays.<NativeModule>asList(
          new TestModuleModule(reactContext)
      );
    }

    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        // Register your native component\'s view manager
        // https://facebook.github.io/react-native/docs/native-components-android.html#4-register-the-viewmanager
        return Arrays.<ViewManager>asList(
            new TestModuleManager()
        );
    }
}
"
`;

exports[`Android/Java: Native Modules creates a TemplateModule.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import android.support.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;

public class TestModuleModule extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = \"TestModule\";
    private static ReactApplicationContext reactContext = null;

    public TestModuleModule(ReactApplicationContext context) {
        // Pass in the context to the constructor and save it so you can emit events
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        super(context);

        reactContext = context;
    }

    @Override
    public String getName() {
        // Tell React the name of the module
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        return REACT_CLASS;
    }

    @Override
    public Map<String, Object> getConstants() {
        // Export any constants to be used in your native module
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
        final Map<String, Object> constants = new HashMap<>();
        constants.put(\"EXAMPLE_CONSTANT\", \"example\");

        return constants;
    }

    @ReactMethod
    public void exampleMethod () {
        // An example native method that you will expose to React
        // https://facebook.github.io/react-native/docs/native-modules-android.html#the-toast-module
    }

    private static void emitDeviceEvent(String eventName, @Nullable WritableMap eventData) {
        // A method for emitting from the native side to JS
        // https://facebook.github.io/react-native/docs/native-modules-android.html#sending-events-to-javascript
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName, eventData);
    }
}
"
`;

exports[`Android/Java: Native Modules creates a TemplatePackage.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class TestModulePackage implements ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
      // Register your native module
      // https://facebook.github.io/react-native/docs/native-modules-android.html#register-the-module
      return Arrays.<NativeModule>asList(
          new TestModuleModule(reactContext)
      );
    }

    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }
}
"
`;

exports[`Android/Java: UI Components creates a TemplateManager.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import android.view.View;

import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;

import com.facebook.react.uimanager.annotations.ReactProp;

public class TestModuleManager extends SimpleViewManager<View> {
    public static final String REACT_CLASS = \"TestModule\";

    @Override
    public String getName() {
        // Tell React the name of the module
        // https://facebook.github.io/react-native/docs/native-components-android.html#1-create-the-viewmanager-subclass
        return REACT_CLASS;
    }

    @Override
    public View createViewInstance(ThemedReactContext context){
        // Create a view here
        // https://facebook.github.io/react-native/docs/native-components-android.html#2-implement-method-createviewinstance
        return new View(context);
    }

    @ReactProp(name = \"exampleProp\")
    public void setExampleProp(View view, String prop) {
        // Set properties from React onto your native component via a setter method
        // https://facebook.github.io/react-native/docs/native-components-android.html#3-expose-view-property-setters-using-reactprop-or-reactpropgroup-annotation
    }
}
"
`;

exports[`Android/Java: UI Components creates a TemplatePackage.java 1`] = `
"//  Created by react-native-create-bridge

package com.testapp.testmodule;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class TestModulePackage implements ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }

    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        // Register your native component\'s view manager
        // https://facebook.github.io/react-native/docs/native-components-android.html#4-register-the-viewmanager
        return Arrays.<ViewManager>asList(
            new TestModuleManager()
        );
    }
}
"
`;
